### PR TITLE
[Core] Clean up constraints classes. Making member variables of linear constraint protected

### PR DESCRIPTION
**Pull request by loumalouomega**
the original poupose of this PR was to move the member variables of the linear_master_slave_constraint to protected, in order to derive new constraints classes. I decided additionally to clean up and extend documentation

### CODE REVIEW

File: kratos/includes/linear_master_slave_constraint.h [hunk 1]
```diff
@@ -135,132 +211,248 @@ class LinearMasterSlaveConstraint :  public MasterSlaveConstraint
...
     {
-        this->SetId(rOther.Id());
-        // this->Flags = rOther.Flags;
-        this->mSlaveDofsVector = rOther.mSlaveDofsVector;
-        this->mMasterDofsVector = rOther.mMasterDofsVector;
-        this->mRelationMatrix = rOther.mRelationMatrix;
-        this->mConstantVector = rOther.mConstantVector;
-
+        BaseType::operator=( rOther );
+        mSlaveDofsVector = mSlaveDofsVector;
```
Comment (adityaghantasala) [inline kratos/includes/linear_master_slave_constraint.h]: this should be 
` mSlaveDofsVector = rOther.mSlaveDofsVector;` or ?

also in the following lines

**Review by adityaghantasala — CHANGES_REQUESTED** [review]

**Comment by loumalouomega** [comment]
True, solved @adityaghantasala

**Comment by adityaghantasala** [comment]
Approving ! Thanks for putting in the doxygen comments !
One question though : SetData and GetData why do this base class need them ?

**Review by adityaghantasala — APPROVED** [review]

**Comment by loumalouomega** [comment]
To access to the data container, the clone method uses this

**Pull request merged**